File: src/lessons/maze/IslandMazeEntity.java
package lessons.maze;

import jlm.universe.Direction;

public class IslandMazeEntity extends jlm.universe.bugglequest.SimpleBuggle {
	@Override
	public void setX(int i)  {
		if (isInited())
			throw new RuntimeException("setX(int) forbidden in this exercise");
	}
	@Override
	public void setY(int i)  { 
		if (isInited())
			throw new RuntimeException("setY(int) forbidden in this exercise");
	}
	@Override
	public void setPos(int i,int j)  { 
		if (isInited())
			throw new RuntimeException("setPos(int,int) forbidden in this exercise");
	}

	/* BEGIN SOLUTION */
	 public void run() {
		 int state = 0 ;
		 this.setDirection(this.chosenDirection);
		 while ( !isOverBaggle() )
		 {
			 switch ( state )
			 {
			 case 0: // North runner mode
				 while ( !isFacingWall() )
				 {
					 forward();
			     }
				 this.turnRight(); // make sure that we have a left wall
				 state = 1; // time to enter the Left Follower mode
			 break;
			 case 1: // Left Follower Mode
				 this.stepHandOnWall(); // follow the left wall
		         if ( this.isChosenDirectionFree() && (this.getDirection() == this.chosenDirection)  ) 
		         {
		        	 state =0; // time to enter in North Runner mode
			     }
				          break;
		      }
		 }
		 this.pickUpBaggle();
	 }

	private void stepHandOnWall(){
		while ( ! isFacingWall() )
		{
			forward();
			turnLeft();
		}
		turnRight();
	}

	Direction chosenDirection = Direction.NORTH;

	private boolean isChosenDirectionFree() {
		Direction memorizedD = getDirection();
		boolean isFree = false;
		this.setDirection(this.chosenDirection);
		if (!isFacingWall()) {
			isFree=true;
		} 
		this.setDirection(memorizedD);
		return isFree;
	}
	

	/* END TEMPLATE */
	
}

File: src/lessons/maze/island/IslandMazeEntity.py
def setX(i):
        errorMsg("Sorry Dave, I'm afraid I cannot let you use setX(i) in this exercise")
def setY(i):
        errorMsg("Sorry Dave, I'm afraid I cannot let you use setY(i) in this exercise")
def setPos(x,y):
        errorMsg("Sorry Dave, I'm afraid I cannot let you use setPos(x,y) in this exercise")


# BEGIN SOLUTION 
def isDirectionFree(dir):
    memo = getDirection()
    setDirection(dir)
    res = not isFacingWall()
    setDirection(memo)
    return res

def stepHandOnWall():
    # PRE: we have a wall on the left
    # POST: we still have the same wall on the left, are one step ahead
    while not isFacingWall():
        forward()
        left() # change to right to get a right follower
    right() # change to left to get a right follower

northRunner = True
chosenDir = Direction.NORTH
setDirection(chosenDir)

while not isOverBaggle():
    if northRunner:
        while not isFacingWall():
            forward()
        right()
        northRunner = False
    else: # left follower mode
        stepHandOnWall()
        if isDirectionFree(chosenDir) and getDirection() == chosenDir:
            northRunner = True

pickupBaggle()

# END SOLUTION
